mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/MapperContext.java

```diff
@@ -18,87 +18,48 @@
-
-  public MapperContext(CqlSession session) {
-    this(session, null, null, new ConcurrentHashMap<>());
-  }
-
-  private MapperContext(
-      CqlSession session,
-      CqlIdentifier keyspaceId,
-      CqlIdentifier tableId,
-      ConcurrentMap<Class<? extends NameConverter>, NameConverter> nameConverterCache) {
-    this.session = session;
-    this.keyspaceId = keyspaceId;
-    this.tableId = tableId;
-    this.nameConverterCache = nameConverterCache;
-  }
-
-  public MapperContext withKeyspaceAndTable(CqlIdentifier newKeyspaceId, CqlIdentifier newTableId) {
-    return (Objects.equals(newKeyspaceId, this.keyspaceId)
-            && Objects.equals(newTableId, this.tableId))
-        ? this
-        : new MapperContext(session, newKeyspaceId, newTableId, nameConverterCache);
-  }
-
-  public CqlSession getSession() {
-    return session;
-  }
+  @NonNull
+  CqlSession getSession();
 
   /**
    * If this context belongs to a DAO that was built with a keyspace-parameterized mapper method,
    * the value of that parameter. Otherwise null.
    */
-  public CqlIdentifier getKeyspaceId() {
-    return keyspaceId;
-  }
+  @Nullable
+  CqlIdentifier getKeyspaceId();
 
   /**
    * If this context belongs to a DAO that was built with a table-parameterized mapper method, the
    * value of that parameter. Otherwise null.
    */
-  public CqlIdentifier getTableId() {
-    return tableId;
-  }
+  @Nullable
+  CqlIdentifier getTableId();
 
   /**
    * Returns an instance of the given converter class.
    *
    * <p>The results of this method are cached at the mapper level. If no instance of this class
    * exists yet for this mapper, a new instance is built by looking for a public no-arg constructor.
    */
-  public NameConverter getNameConverter(Class<? extends NameConverter> converterClass) {
-    return nameConverterCache.computeIfAbsent(converterClass, MapperContext::buildNameConverter);
-  }
+  @NonNull
+  NameConverter getNameConverter(Class<? extends NameConverter> converterClass);
 
-  private static NameConverter buildNameConverter(Class<? extends NameConverter> converterClass) {
-    try {
-      return converterClass.getDeclaredConstructor().newInstance();
-    } catch (InstantiationException
-        | IllegalAccessException
-        | NoSuchMethodException
-        | InvocationTargetException e) {
-      throw new IllegalStateException(
-          String.format(
-              "Error while building an instance of %s. "
-                  + "%s implementations must have a public no-arg constructor",
-              converterClass, NameConverter.class.getSimpleName()),
-          e);
-    }
-  }
+  /**
+   * Retrieves any custom state that was set while building the mapper with {@link
+   * MapperBuilder#withCustomState(Object, Object)}.
+   *
+   * <p>The returned map is immutable. If no state was set on the builder, it will be empty.
+   */
+  @NonNull
+  Map<Object, Object> getCustomState();
```

Comment:
Not sure we need to expose the entire map. The typical API for context objects is `context.getAttribute(String name)` and `context.setAttribute(String name, Object value)`, and the underlying map implementation is completely hidden. E.g. Jackson uses this convention in its `DeserializationContext`.

Reply:
But then you lose the ability to iterate keys. As I said this feature should be rarely needed, let's not overthink it.